### DISCUSSION TITLE
Support for Windows and MacOs apps development

### DISCUSSION THREAD

**Discussion by firasrg** [expo/expo#22273] [Proposals and Ideas]
Hello!

I've read on React Native website something interesting! it seems that Microsoft has made some libraries to empower React Native, in order to make it suitable to [build Windows/MacOS apps](https://microsoft.github.io/react-native-windows/).

Now I think about **Expo**, what is the plan ?? is it possible for Expo to build for Windows/MacOS like it does for mobile/web now ?

I'd like to open this topic and listen to community thoughts, including Expo team.

Thank you for your attention.

**Comment by stephenhandley** [comment]
Yeah there's also [react-native-macos](https://github.com/microsoft/react-native-macos). Targeting desktops would be great, but sounds like [it is currently not a priority](https://docs.expo.dev/workflow/prebuild/#platform-compatibility). 

This looks to be deprecated, but suggests running the expo web target via Electron:
https://github.com/expo/expo-electron-adapter

At least ChatGPT thinks it's possible 🤷‍♂️ 
<img width="689" alt="Screenshot 2023-07-28 at 14 50 46" src="https://github.com/expo/expo/assets/3257/a69ced8f-5758-447d-93a9-bb943e6f5722">

(it isn't 👻)  
```
$ expo add:platform macos
"add:platform,macos" is not an expo command. See "expo --help" for the full list of commands.
```

**Comment by tomByrer** [comment]
Seems Expo team is not too interested in Win/MacOS desktop platforms, so perhaps the discussion can be now,
"How to use both Expo & MS-RN4D in the same project with max codeshare?"

**Reply by tsapeta** [reply]
We don't have any plans to support Windows at the moment, but we're looking into supporting macOS. By supporting I mean:
- making it possible to create native Expo Modules for macOS (supporting this platform in the core part)
- enabling macOS platform in some modules (most of them don't need too much platform-specific implementation)

I already did some work for this, but I found some issues in react-native-macos regarding the support for Swift libraries, which hopefully will be fixed in 0.73. Then I'll be able to move forward 😉

**Reply by 0xA-10** [reply]
Is https://github.com/microsoft/react-native-macos/pull/1967 the issue you are referring to? If so, it looks like it might just need some pushing along to be merged.

Thanks for your effort on this!

**Reply by tsapeta** [reply]
It's not related to `use_frameworks!` but pretty close. The issue was fixed in React Native itself and then cherry-picked to react-native-macos as part of the upgrade to 0.73.
I can confirm that it works fine on 0.73 so I'm no longer blocked on this.

I've already merged a few PRs that add support for macOS in the core modules and autolinking. Stay tuned! 😉

**Reply by jtwebman** [reply]
Any thing we can help out with? I would love to take a stab at helping on the windows side as well.

**Reply by bretthayes** [reply]
Any update on the macOS support @tsapeta? I've been keeping an eye on your related [closed PRs](https://github.com/expo/expo/pulls?q=is%3Apr+macos+author%3Atsapeta+is%3Aclosed+). Would love to see something like this in the near future:

```json
"scripts": {
  "start": "expo start",
  "aos": "expo start --aos",
  "ios": "expo start --ios",
  "osx": "expo start —osx",
  "win": "expo start —win"
}
```

The 3 letter acronyms look clean 😉

**Reply by serafdev** [reply]
What if Expo, Flutter, Electron merged together to solve the cross-platform once and for all 😆

**Reply by jtwebman** [reply]
I think no matter what you will always have a different project for mobile vs windows vs mac as you are still running native so you are going to have different UI to build for each. I am ok with this as I want it to integrate the best it can into the OS to support all the features I can.

**Reply by serafdev** [reply]
> I think no matter what you will always have a different project for mobile vs windows vs mac as you are still running native so you are going to have different UI to build for each. I am ok with this as I want it to integrate the best it can into the OS to support all the features I can.

What if the native just becomes the linux kernel @jtwebman so we stop wasting everyone's time

**Reply by jtwebman** [reply]
We can have dreams for sure!

**Comment by Andriiklymiuk** [comment]
well, currently there is this issue, if you add macOS to current expo project (ExpoModulesCode seems to be failing build somehow)
https://github.com/facebook/react-native/issues/32704
maybe someone have figured out how to add macOS support to current expo? 

trying to find workaround myself, but not sure it will be successful

**Comment by srehan17** [comment]
I am trying to convert a React Native Expo app to React Native Windows app and React Native MacOC app. To test if that is possible, I created a project using React Native Expo, ran npx expo prebuild to generate the native source code, and then tried to add macOS/Windows app dependencies.

On macOS, I am facing the ExpoModulesCore issue. 

On Windows, I am facing fmt issue:
\node_modules\.fmt\fmt-8.0.0\include\fmt\format.h(341,51): error C4996: 'stdext::checked_array_iterator<T*>': warning STL4043: stdext::checked_array_iterator, stdext::unchecked_array_iterator, and related factory functions are non-Standard extensions and will be removed in the future. std::span (since C++20) and gsl::span can be used instead. You can define _SILENCE_STDEXT_ARR_ITERS_DEPRECATION_WARNING or _SILENCE_ALL_MS_EXT_DEPRECATION_WARNINGS to suppress this warning. 

Any suggestions / workarounds?

**Comment by luchillo17** [comment]
At this point, it looks like the next major player in the cross-platform app will be Tauri, it's not there yet but they intend to support Web, Mobile, and Desktop, so instead of grabbing the Expo app and wrapping the web version in Electron, just using Tauri directly seems simpler and more performant (at least in desktop).

**Reply by isaachinman** [reply]
The problem with Tauri is that it is no better than Capacitor – it uses the OS webview. On iOS and macOS, this is a massive pain, as WKWebView is horrible.

That is the entire draw of Electron: a stable runtime regardless of platform.

I do believe that the only reasonable options are: (1) Electron, or (2) actual native UI components. I would expect a RN-to-desktop implementation to target the second option.

**Reply by code-504** [reply]
Yeah, it is a drawback using Tauri. It's easy to set up along with a React Native project, but it has inconsistencies among platforms, and you also need to learn Rust to fully configure Tauri.

**Comment by brawler2011** [comment]
So is there current guideline for developing an app for win/macos with expo?

**Comment by kirill-konshin** [comment]
Here's the screenshot from recent React Conf 2024

<img width="1345" alt="image" src="https://github.com/expo/expo/assets/1530410/91792c81-238b-4264-bcc5-a8255d0e7076">

@EvanBacon please shed some light on building for Mac? And possibly on Windows using RN for Windows?

**Reply by tsapeta** [reply]
This was a Mac Catalyst app – a macOS version of the app for iPads. It's not using `react-native-macos`.

**Reply by kirill-konshin** [reply]
Thanks @tsapeta, yep, the one that works on M Macs. But I can't find instructions how to build an Expo app for use with Catalyst... are there any extra steps needed?

Apparently we already had a similar discussion in 2021 https://github.com/expo/expo/issues/7166#issuecomment-761745798 :)

**Reply by kirill-konshin** [reply]
@tsapeta @EvanBacon it would be great to get instructions how to build Expo apps for Catalyst. I can't find any documentation that explains this...

**Comment by buildnewapp** [comment]
Continuous attention
I don't want to export the Expo to the web version and then package it as a client using Electron or Tauri.

**Comment by anasvemmully** [comment]
any update...It would be amazing to have support window/mac version! :)

**Comment by jenseralmeida** [comment]
Is this what the https://github.com/expo/expo/pull/30309 is all about, but For MACOS only?

**Comment by erfanazari** [comment]
So i did it and worked for me:
first you need to install visual studio.
then run `npx expo eject` on your expo app.
then install react-native-windows using `npm install --save react-native-windows@latest`.
then initialize it using `npx react-native init-windows --overwrite`.
now you can run it using `npx react-native run-windows` or `npx react-native autolink-windows`.
That's it!

**Reply by jtwebman** [reply]
Does this also work in WSL2 in Ubuntu or do you need to be in the windows system?

**Reply by erfanazari** [reply]
I haven't tested it on ubuntu or WSL2.
But i know that it works on windows

**Reply by code-504** [reply]
Have you tested it on Expo Web?

I was able to use it on Android, iOS, and Windows, but when I tried to configure it for Web, Expo on Windows stopped working.

**Reply by Sowed** [reply]
Have you tried bundling and making a release build for your app @erfanazari?

**Comment by FatahChan** [comment]
been a while, is there any updates on this?

**Comment by omerhac** [comment]
Expo for desktop windows builds wanted! :)

**Comment by hhyyperion** [comment]
https://docs.expo.dev/modules/additional-platform-support
Should get you what you need (for mac)

**Reply by Arjan-Zuidema** [reply]
That's just for building a module, not using react-native-macos in expo

**Comment by AlexSubZero** [comment]
Any update on if expo will support windows?

**Comment by chriszs** [comment]
It's frustrating to watch PRs that would make solid, incremental progress on enabling support for this languish while others sail through without review. It feels like the roadmap at Expo is being driven by a specific team, which doesn't believe Mac development should be part of the agenda. But I think it should be. I interact with a lot of Electron apps daily, which can and do share a lot of code with the web versions of those apps, but which are not native experiences, and take up more resources than they really should. I developed one of these apps, because I had to prioritize desktop, and right now that means Electron, even though my preference in a stack would be React Native/Expo to set up a cross-platform app, which is what I do on another project. It's true that the Mac development experience, even if you solved the Expo part of it, would be rough, but there's a chicken and egg problem that is actively getting solved for the web platform because of Expo's support for it. If React Native's recommendation  is "use a framework" and Expo is the only currently recommended framework, then our solid alternatives are... Rock (formerly RNEF)? Stick Expo web in Electron?

**Reply by isaachinman** [reply]
> Stick Expo web in Electron

This is currently the only good option, as far as I can tell

**Comment by a-eid** [comment]
any updates on this ?

**Comment by mempler** [comment]
Does anyone know an alternative to expo except the react native cli? Given that expo seemingly to refuse to implement it.

I'm willing to fully drop expo in favor of another product to spend some money in just to get proper native desktop support.

**Reply by isaachinman** [reply]
Expo Web + Electron is probably still your best bet.

**Reply by mempler** [reply]
Yea... no, not a viable alternative. _native_ is the key here

**Reply by isaachinman** [reply]
Then bang your head against `react-native-macos`.

There are no other options that don't involve webviews, unless you consider Flutter, which is a joke.

**Reply by stephenhandley** [reply]
https://v2.tauri.app/

**Reply by isaachinman** [reply]
> don't involve webviews

Save yourself the headache

**Comment by abdymm** [comment]
do you guys have some proper docs on how to integrate Expo with Electron
Ive successfully built it, but when try to run it, it just like opening the static `index.html` file from `dist` (build) , so no js running like when we run it thru web

if you guys have experience building this please share it hereeee 🙌

**Reply by isaachinman** [reply]
```ts
const loadURL = serve({
  directory: path.join(__dirname, '..', 'dist'),
})
```

Just do some basic research. It's literally boilerplate Electron.

**Comment by litianming1** [comment]
any updates on this ?

**Comment by chriszs** [comment]
This is the top voted discussion topic, which seems like a signal, but the votes pale in comparison to some of the other feature requests in [Canny](https://expo.canny.io), which is officially where feature requests are supposed to live. I looked for a React Native MacOS feature request and didn't find one (though I found Marzipan/Catalyst and RN Windows), so I made one: https://expo.canny.io/feature-requests/p/support-macos-apps-with-react-native-macos

**Comment by tsapeta** [comment]
As someone from the Expo team I feel that I should chime in here and dispel some doubts because this discussion is getting popular. We've already put a lot of efforts into supporting the macOS platform, but at the moment maintaining this platform would be too expensive for us. Please note that the latest version of `react-native-macos` is 0.79 but in our monorepo we need to use the latest `react-native` version 0.83 (rc) and nightlies of 0.84. This is kind of a blocker for us, because it basically means we will have to support much more versions of React Native backwards. As long as `react-native-macos` doesn't catch up with upgrades, we can't invest all our resources in this. Perhaps using Mac Catalyst would be easier and more straightforward? Definitely something to consider.

We're trying to keep a few of Expo modules working on macOS and are open for any pull requests adding support for macOS in any other module from the SDK though. There is a guide in the official docs [how to install Expo modules](https://microsoft.github.io/react-native-macos/docs/guides/installing-expo-modules).
Sadly, building the first-class support for macOS is a much larger scope of work.

One of my ideas how to improve the current state is to decouple iOS-specific implementation in React Native. Then it's easy to replace it with implementations for other platforms (potentially also SwiftUI-only targets). Once done, `react-native-macos` doesn't have to be a fork thus keeping up with upgrades would be much easier as the codebase would be much smaller and platform-specific. In practice it should be more resistant to breaking and internal changes in React Native. We're discussing with the React Native team about other platforms very frequently (I even shared this idea just a week ago while I was visiting them in London) although this will certainly not be a priority before the release of the hypothetical version 1.0.

Emphasizing what I wrote above, it does not mean we're abandoning this idea, it's just not the highest priority and we can put only limited resources into this.

**Reply by mempler** [reply]
For me, macOS can be entirely scrapped, Mac Catalyst is good enough. The biggest hurdle is Windows, really. Especially that's where most of the end-users are gonna be.

Unfortunately though, they're too tightly coupled. And the only alternative is an horrible memory hog (Electron, and it's horrible brother, WebView2)

I don't think there is a need for first class support though, only "It works", every other component can be maintained by the community.

**Reply by isaachinman** [reply]
Unfortunately, on the contrary, I've found Mac Catalyst to be a complete joke for data-heavy apps.

**Reply by litianming1** [reply]
When the Windows/MacOS SDK versions catch up, will support become possible?

**Comment by a-eid** [comment]
I think now that AI is everywhere I think expo should try to support desktop as well.

**Comment by alexoloopios** [comment]
I've been looking into React Native on desktop and honestly it would be better using native components than things like Electron and webviews.

**Comment by jerryseigle** [comment]
Mac Catalyst  is good for me can anyone share any docs to how to make a Mac Catalyst build so far Mac Catalyst  won't build the default Expo app

I get this error
Command PhaseScriptExecution failed with a nonzero exit code

**Comment by oliveryasuna** [comment]
Support for Electrobun would be great too!

**Reply by a-eid** [reply]
`Electrobun` is not native.

**Reply by oliveryasuna** [reply]
> `Electrobun` is not native.

True, but why does this matter? There is still an API for interacting with native code.

**Comment by eveningkid** [comment]
You might want to take a look at @shirakaba's [expo-desktop](https://github.com/shirakaba/expo-desktop)
